README.md

```diff
@@ -54,10 +54,29 @@ module "vpc" {
   name = "my-vpc"
   cidr = "10.0.0.0/16"
 
-  azs             = ["eu-west-1a", "eu-west-1b", "eu-west-1c"]
-  private_subnets = ["10.0.1.0/24", "10.0.2.0/24", "10.0.3.0/24"]
-  public_subnets  = ["10.0.101.0/24", "10.0.102.0/24", "10.0.103.0/24"]
+  private_subnets = {
+    "subnet-1" = {
+      cidr = "10.0.1.0/24",
+      az   = "eu-west-1a"
+    },
+    "subnet-2" = {
+      cidr = "10.0.2.0/24",
+      az   = "eu-west-1a"
+    }
```

Comment:
@antonbabenko I didn't test it for now, but Is this structure of the subnets variable ok for you ?

Reply:
The structure is fine, but I am thinking about an important note on the implementation - aim to have backward compatibility (type `list(string)` should be converted into `map(map(any)`). Have `type = any` on variables.

We can't just ask users to rewrite their code at once even if we release a new major version.

Reply:
> The structure is fine, but I am thinking about an important note on the implementation - aim to have backward compatibility (type list(string) should be converted into map(map(any)). Have type = any on variables.

Are you proposing to support for both `list(string)` and `map(map(string))` inputs ? Or to help users to move as easy as possible to the new the input structure (`map(map(string))`) ?

> We can't just ask users to rewrite their code at once even if we release a new major version.

I think we have to document it. We can't do more without adding an extra tool that we should maintain. We can probably provide something for states manipulation, but I'm not sure we want to go through terraform code generation.

Reply:
1. `type = any` (since `type = list(string)|map(map(any))` or similar is not supported). Yes, support both types of values transparently for the users.
2. Agree. We have to document but first, let's try to make it backward compatible. Code generation or even relying on tools like [tfmigrate](https://github.com/minamijoyo/tfmigrate) is not the best option for such modules to my mind.

Reply:
> 1. `type = any` (since `type = list(string)|map(map(any))` or similar is not supported). Yes, support both types of values transparently for the users.

I was thinking about using `map(map(string))` and not `map(map(any))`. But using `any` is also ok for me.

> 2. Agree. We have to document but first, let's try to make it backward compatible. Code generation or even relying on tools like [tfmigrate](https://github.com/minamijoyo/tfmigrate) is not the best option for such modules to my mind.

Can you please elaborate your point ? How do you want to handle this backward compatibility without a lot of complexity ? If you have something in mind, please share it.

Reply:
1. `map(map(any))` is better for extending a list of subnets with support for tags (which can be `list(any)` and not `string`). For eg:
```
public_subnet = {
"one" = {
    cidr = "10.0.1.0/24"
    az   = "eu-west-1a"
    tags = ["some", "thing"]
}
...
}
```

2. I have some thoughts, and I will try to put them in writing before our meeting on Friday.

Reply:
@barryib 

```terraform
locals {
  one_map = try(tomap(var.subnets), {for k in var.subnets: k => {}})
}

variable "subnets" {
  type = any
  default = ["10.0.0.0/20", "10.0.1.0/20"]
#  default = {
#    "10.0.0.0/20" = { az = "eu-west-1a", tags = { Name: "first" }}
#    "10.0.1.0/20" = { az = "eu-west-1b", tags = { Name: "second" }}
#  }
}
```

`var.subnets` can be list or map, and Terraform will convert that into a map. See around this time - https://youtu.be/xpfGBgLgpWo?t=1520

Reply:
Humm.. sounds nice. I'll try that. But this still involve state manipulation right ? It just help users to not modify their code base.

Should we use this for transition only (said for 1 or 2 major release) and we invite them to change their code ?

IMHO, I think that users will be more concerne about state migration and to avoid rebuilding all their infrastructure. The minor code change can be acceptable.

Reply:
Yes, manual state manipulation will probably still required, unless we can generate keys to match indexes in a list, like this (to use maps but keep old indexes):

```
one_map = try(tomap(var.subnets), {for k,v in var.subnets: k => { cidr = v }})
```

I think if we use the type conversion, we can keep using it for a long time (more than 1-2 major releases) to support both cases transparently. But we should also see if it is going to look ugly and hard to maintain, we can drop support for older structure (list).

Reply:
> Yes, manual state manipulation will probably still required, unless we can generate keys to match indexes in a list, like this (to use maps but keep old indexes):

That was the first thing I tried, but I'm afraid that it won't probably work as expected (but we can still test it). As you mention, map keys are strings and you can't change their type (unless you have an idea).

```shell
# random_pet.node_groups will be destroyed
  - resource "random_pet" "node_groups" {
      - id        = "wondrous-honeybee" -> null
      - keepers   = {
          - "foo" = "10.0.0.0/20"
        } -> null
      - length    = 2 -> null
      - separator = "-" -> null
    }

  # random_pet.node_groups[1] will be destroyed
  - resource "random_pet" "node_groups" {
      - id        = "topical-grizzly" -> null
      - keepers   = {
          - "foo" = "10.0.1.0/20"
        } -> null
      - length    = 2 -> null
      - separator = "-" -> null
    }

  # random_pet.node_groups["0"] will be created
  + resource "random_pet" "node_groups" {
      + id        = (known after apply)
      + keepers   = {
          + "foo" = "10.0.0.0/20"
        }
      + length    = 2
      + separator = "-"
    }

  # random_pet.node_groups["1"] will be created
  + resource "random_pet" "node_groups" {
      + id        = (known after apply)
      + keepers   = {
          + "foo" = "10.0.1.0/20"
        }
      + length    = 2
      + separator = "-"
    }
```

`random_pet.node_groups` => `random_pet.node_groups["0"]`
`random_pet.node_groups[1]` => `random_pet.node_groups["1"]`

Please find bellow my random_pet snippet:

1. I run it first with a `count`
2. I used the magic to convert it to map and used with `for_each`

```hcl
resource "random_pet" "node_groups" {
  for_each = local.one_map
  #count     = length(var.list_subnets)
  separator = "-"
  length    = 2

  keepers = {
    foo = each.value.cidr
    #foo = var.list_subnets[count.index]
  }
}
```

Reply:
No surprise here then :) Good that you double-checked it.